File: python/tables/text/table_text.py
import sys, os, ctypes
import os, sys
# the wrapper lives at <root>/wrappers/python (source checkout) or beside
# the example tree (shipped package) -- find it without a hard-coded path
_d = os.path.dirname(os.path.abspath(__file__))
for _ in range(6):
    for _c in (os.path.join(_d, 'wrappers', 'python'), _d):
        if os.path.isfile(os.path.join(_c, 'lumaspdf.py')):
            sys.path.insert(0, _c)
            break
    else:
        _d = os.path.dirname(_d)
        continue
    break
import lumaspdf as L

# table_text -- port of examples\Vb6\tables\text\table_text.bas
# Builds a 3x3 table demonstrating cell text alignment, draws it, then redraws
# it with a 90-degree cell orientation. Uses the flat tbl* exports.

HERE = os.path.dirname(os.path.abspath(__file__))

ERRPROC = ctypes.WINFUNCTYPE(ctypes.c_int32, ctypes.c_void_p, ctypes.c_int32,
                             ctypes.c_char_p, ctypes.c_int32)


def _err(data, code, msg, typ):
    if msg:
        print(msg.decode("latin-1", "replace"))
    return 0


_cb = ERRPROC(_err)


def _ptr_to_ansi(p):
    if not p:
        return ""
    return ctypes.cast(p, ctypes.c_char_p).value.decode("latin-1", "replace")


def main():
    import time
    t0 = time.perf_counter()

    pdf = L.pdfNewPDF()
    L.pdfSetOnErrorProc(pdf, 0, _cb)

    L.pdfCreateNewPDFA(pdf, b"")
    L.pdfSetPageCoords(pdf, L.pcTopDown)

    tbl = L.tblCreateTable(pdf, 3, 3, 500.0, 100.0)
    L.tblSetBoxProperty(tbl, -1, -1, L.tbpBorderWidth, 1.0, 1.0, 1.0, 1.0)
    L.tblSetFontA(tbl, -1, -1, b"Arial", L.fsRegular, 1, L.cp1252)
    L.tblSetFontA(tbl, -1, 1, b"Arial", L.fsBold, 1, L.cp1252)
    L.tblSetGridWidth(tbl, 1.0, 1.0)

    txt = b"The cell alignment can be set for text, images, and templates..."

    # -1.0 means use the default row height as specified in CreateTable().
    rowNum = L.tblAddRow(tbl, -1.0)
    L.tblSetCellTextA(tbl, rowNum, 0, L.taLeft, L.coTop, txt, len(txt))
    L.tblSetCellTextA(tbl, rowNum, 1, L.taCenter, L.coTop, txt, len(txt))
    L.tblSetCellTextA(tbl, rowNum, 2, L.taRight, L.coTop, txt, len(txt))

    rowNum = L.tblAddRow(tbl, -1.0)
    L.tblSetCellTextA(tbl, rowNum, 0, L.taLeft, L.coCenter, txt, len(txt))
    L.tblSetCellTextA(tbl, rowNum, 1, L.taCenter, L.coCenter, txt, len(txt))
    L.tblSetCellTextA(tbl, rowNum, 2, L.taRight, L.coCenter, txt, len(txt))

    rowNum = L.tblAddRow(tbl, -1.0)
    L.tblSetCellTextA(tbl, rowNum, 0, L.taLeft, L.coBottom, txt, len(txt))
    L.tblSetCellTextA(tbl, rowNum, 1, L.taCenter, L.coBottom, txt, len(txt))
    L.tblSetCellTextA(tbl, rowNum, 2, L.taRight, L.coBottom, txt, len(txt))

    # Draw the table now
    L.pdfAppend(pdf)
    L.tblDrawTable(tbl, 50.0, 50.0, 742.0)
    while L.tblHaveMore(tbl) != 0:
        L.pdfEndPage(pdf)
        L.pdfAppend(pdf)
        L.tblDrawTable(tbl, 50.0, 50.0, 742.0)
    L.pdfEndPage(pdf)

    # Change the cell orientation to see what happens...
    L.tblSetCellOrientation(tbl, -1, -1, 90)
    L.pdfAppend(pdf)
    L.pdfSetFontA(pdf, b"Arial", L.fsRegular, 12.0, 1, L.cp1252)
    L.pdfWriteTextA(pdf, 50.0, 50.0,
                    b"The same table but the cell orientation was changed to 90 degrees.")

    L.tblDrawTable(tbl, 50.0, 65.0, 742.0)
    while L.tblHaveMore(tbl) != 0:
        L.pdfEndPage(pdf)
        L.pdfAppend(pdf)
        L.tblDrawTable(tbl, 50.0, 50.0, 737.0)
    L.pdfEndPage(pdf)

    _th = ctypes.c_void_p(tbl)
    L.tblDeleteTable(ctypes.byref(_th))    # frees the table

    # A table stores errors and warnings in the error log
    err = L.TPDFError()
    err.StructSize = ctypes.sizeof(err)
    for i in range(L.pdfGetErrLogMessageCount(pdf)):
        L.pdfGetErrLogMessage(pdf, i, ctypes.byref(err))
        print(_ptr_to_ansi(err.Msg))

    # No fatal error occurred?
    if L.pdfHaveOpenDoc(pdf) != 0:
        out_file = os.path.join(HERE, "out.pdf")
        if L.pdfOpenOutputFileA(pdf, out_file.encode("latin-1")) == 0:
            L.pdfDeletePDF(pdf)
            return
        if L.pdfCloseFile(pdf) != 0:
            print("Processing time: %d ms" % int((time.perf_counter() - t0) * 1000))

    L.pdfDeletePDF(pdf)


if __name__ == "__main__":
    main()

File: vbnet/tables/text/table_text.vb
' table_text -- VB.NET port of examples\Vb6\tables\text\table_text.bas
' Builds a 3x3 table demonstrating cell text alignment (left/center/right x
' top/center/bottom), draws it, then redraws it with a 90-degree cell
' orientation.
Imports System
Imports System.Runtime.InteropServices
Imports LumasPdfSdk

Module modTableText
    Private ErrDelegate As TErrorProc
    ' NOTE: the VB6 original passes Len = -1 (the DynaPDF "null-terminated" sentinel).
    ' The LumasPdf engine's flat tblSetCellTextA (Lumas.Pdf.Api.pas) does NOT honour that
    ' sentinel: "if Len > 0 then SetLength(R, Len); Move(Text^, R[1], Len)" treats
    ' &HFFFFFFFF as a ~4GB length and faults with an AccessViolation. Only Len = 0 selects
    ' null-terminated. We therefore pass the real byte length here (equivalent output).
    Private TxtLen As UInteger

    Function AppPath() As String
        Return AppDomain.CurrentDomain.BaseDirectory.TrimEnd("\"c)
    End Function

    Function PDFError(ByVal Data As IntPtr, ByVal ErrCode As Integer, ByVal ErrMessage As String, ByVal ErrType As Integer) As Integer
        Console.WriteLine(ErrMessage)
        Return 0
    End Function

    Sub Main()
        Dim pdf As IntPtr, tbl As IntPtr
        Dim outFile As String, txt As String
        Dim timeStart As Long
        Dim i As Integer, rowNum As Integer
        Dim err As TPDFError

        timeStart = Environment.TickCount

        pdf = LumasPdf.pdfNewPDF()
        ErrDelegate = New TErrorProc(AddressOf PDFError)
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, ErrDelegate)

        LumasPdf.pdfCreateNewPDFW(pdf, "")

        LumasPdf.pdfSetPageCoords(pdf, CInt(TPageCoord.pcTopDown))

        tbl = LumasPdf.tblCreateTable(pdf, 3, 3, 500.0, 100.0)
        LumasPdf.tblSetBoxProperty(tbl, -1, -1, TTableBoxProperty.tbpBorderWidth, 1.0, 1.0, 1.0, 1.0)
        LumasPdf.tblSetFontA(tbl, -1, -1, "Arial", LumasPdfConsts.fsRegular, True, TCodepage.cp1252)
        LumasPdf.tblSetFontA(tbl, -1, 1, "Arial", LumasPdfConsts.fsBold, True, TCodepage.cp1252)
        LumasPdf.tblSetGridWidth(tbl, 1.0, 1.0)

        txt = "The cell alignment can be set for text, images, and templates..."
        TxtLen = CUInt(System.Text.Encoding.ASCII.GetByteCount(txt))

        ' -1.0 means use the default row height as specified in the CreateTable() call.
        rowNum = LumasPdf.tblAddRow(tbl, -1.0)
        LumasPdf.tblSetCellTextA(tbl, CUInt(rowNum), 0, LumasPdfConsts.taLeft, TCellAlign.coTop, txt, TxtLen)
        LumasPdf.tblSetCellTextA(tbl, CUInt(rowNum), 1, LumasPdfConsts.taCenter, TCellAlign.coTop, txt, TxtLen)
        LumasPdf.tblSetCellTextA(tbl, CUInt(rowNum), 2, LumasPdfConsts.taRight, TCellAlign.coTop, txt, TxtLen)

        rowNum = LumasPdf.tblAddRow(tbl, -1.0)
        LumasPdf.tblSetCellTextA(tbl, CUInt(rowNum), 0, LumasPdfConsts.taLeft, TCellAlign.coCenter, txt, TxtLen)
        LumasPdf.tblSetCellTextA(tbl, CUInt(rowNum), 1, LumasPdfConsts.taCenter, TCellAlign.coCenter, txt, TxtLen)
        LumasPdf.tblSetCellTextA(tbl, CUInt(rowNum), 2, LumasPdfConsts.taRight, TCellAlign.coCenter, txt, TxtLen)

        rowNum = LumasPdf.tblAddRow(tbl, -1.0)
        LumasPdf.tblSetCellTextA(tbl, CUInt(rowNum), 0, LumasPdfConsts.taLeft, TCellAlign.coBottom, txt, TxtLen)
        LumasPdf.tblSetCellTextA(tbl, CUInt(rowNum), 1, LumasPdfConsts.taCenter, TCellAlign.coBottom, txt, TxtLen)
        LumasPdf.tblSetCellTextA(tbl, CUInt(rowNum), 2, LumasPdfConsts.taRight, TCellAlign.coBottom, txt, TxtLen)

        ' Draw the table now
        LumasPdf.pdfAppend(pdf)
        LumasPdf.tblDrawTable(tbl, 50.0, 50.0, 742.0)
        Do While LumasPdf.tblHaveMore(tbl)
            LumasPdf.pdfEndPage(pdf)
            LumasPdf.pdfAppend(pdf)
            LumasPdf.tblDrawTable(tbl, 50.0, 50.0, 742.0)
        Loop
        LumasPdf.pdfEndPage(pdf)

        ' Let's change the cell orientation to see what happens...
        LumasPdf.tblSetCellOrientation(tbl, -1, -1, 90)
        LumasPdf.pdfAppend(pdf)
        LumasPdf.pdfSetFontW(pdf, "Arial", LumasPdfConsts.fsRegular, 12.0, 1, TCodepage.cp1252)
        LumasPdf.pdfWriteTextW(pdf, 50.0, 50.0, "The same table but the cell orientation was changed to 90 degrees.")

        LumasPdf.tblDrawTable(tbl, 50.0, 65.0, 742.0)
        Do While LumasPdf.tblHaveMore(tbl)
            LumasPdf.pdfEndPage(pdf)
            LumasPdf.pdfAppend(pdf)
            LumasPdf.tblDrawTable(tbl, 50.0, 50.0, 737.0)
        Loop
        LumasPdf.pdfEndPage(pdf)

        LumasPdf.tblDeleteTable(tbl)                  ' frees the table and sets tbl to 0

        ' A table stores errors and warnings in the error log
        err = New TPDFError()
        err.StructSize = CUInt(Marshal.SizeOf(GetType(TPDFError)))
        For i = 0 To LumasPdf.pdfGetErrLogMessageCount(pdf) - 1
            LumasPdf.pdfGetErrLogMessage(pdf, CUInt(i), err)
            Console.WriteLine(Marshal.PtrToStringAnsi(err.Msg))
        Next i

        ' No fatal error occurred?
        If LumasPdf.pdfHaveOpenDoc(pdf) Then
            ' We write the file into the application directory.
            outFile = AppPath() & "\out.pdf"
            If Not LumasPdf.pdfOpenOutputFileW(pdf, outFile) Then
                LumasPdf.pdfDeletePDF(pdf)
                Return
            End If
            If LumasPdf.pdfCloseFile(pdf) Then
                Console.WriteLine("Processing time: " & (Environment.TickCount - timeStart) & " ms")
            End If
        End If

        LumasPdf.pdfDeletePDF(pdf)
    End Sub
End Module
